tests/rkt_non_root_test.go

```diff
@@ -0,0 +1,191 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"runtime"
+	"syscall"
+	"testing"
+
+	"github.com/coreos/rkt/common"
+)
+
+// TestNonRootReadInfo tests that non-root users that can do rkt list, rkt image list.
+func TestNonRootReadInfo(t *testing.T) {
+	ctx := newRktRunCtx()
+	defer ctx.cleanup()
+
+	// Do 'rkt install. and launch some pods in root'.
+	cmd := fmt.Sprintf("%s install", ctx.cmd())
+	t.Logf("Running rkt install")
+	runRktAndCheckOutput(t, cmd, "rkt directory structure successfully created", false)
+
+	// Launch some pods, this creates the environment for later testing,
+	// also it exercises 'rkt install'.
+	imgs := []struct {
+		name     string
+		msg      string
+		exitCode string
+		imgFile  string
+	}{
+		{name: "inspect-1", msg: "foo-1", exitCode: "1"},
+		{name: "inspect-2", msg: "foo-2", exitCode: "2"},
+		//{name: "inspect-3", msg: "foo-3", exitCode: "3"}, // Waiting for #1453 #1503.
+	}
+
+	for i, img := range imgs {
+		imgName := fmt.Sprintf("rkt-%s.aci", img.name)
+		imgs[i].imgFile = patchTestACI(imgName, fmt.Sprintf("--name=%s", img.name), fmt.Sprintf("--exec=/inspect --print-msg=%s --exit-code=%s", img.msg, img.exitCode))
+		defer os.Remove(imgs[i].imgFile)
+	}
+
+	runCmds := []string{
+		// Run with overlay, no private-users.
+		fmt.Sprintf("%s --insecure-skip-verify run --mds-register=false %s", ctx.cmd(), imgs[0].imgFile),
+
+		// Run without overlay, no private-users.
+		fmt.Sprintf("%s --insecure-skip-verify run --no-overlay --mds-register=false %s", ctx.cmd(), imgs[1].imgFile),
+
+		// Run without overlay and private-users.
+		//fmt.Sprintf("%s --insecure-skip-verify run --no-overlay --private-users --mds-register=false %s", ctx.cmd(), imgs[2].imgFile),
+	}
+
+	for i, cmd := range runCmds {
+		t.Logf("#%d: Running %s", i, cmd)
+		runRktAndCheckOutput(t, cmd, imgs[i].msg, false)
+	}
+
+	gid, err := common.LookupGid(common.RktGroup)
+	if err != nil {
+		t.Skipf("Skipping the test because there's no %q group", common.RktGroup)
+	}
+
+	done := make(chan struct{})
+
+	// Go disables setuid/setgid because of its runtime model make
+	// these syscall not effactive in most case because:
+	// 1. Only the kernal thread(not process) will successfully setuid/setgid
+	// after the syscall.
+	// 2. Kernal thread is not binded to goroutine.
+	//
+	// Here we use LockOSThread() and Syscall() to work around this.
+	// Also we need to run in a groutine and destroy the kernel thread on exit
+	// so we are able to clean up the context in root.
+	// For more details, see: https://code.google.com/p/go/issues/detail?id=1435
+	go func() {
+		runtime.LockOSThread()
+		defer func() {
+			close(done)
+			// Destroy the thread so that other goroutines will not run in non-root.
+			syscall.Syscall(syscall.SYS_EXIT, uintptr(0), 0, 0)
```

Comment:
@sgotti Here is an example I run, and I checked the /proc, the thread does exit. 
https://gist.github.com/yifan-gu/af16d1fbafd07d9e8e67

The only problem is there are some chances the exiting thread happens to be the main thread, which can cause the process to be in zombie state.

Other interesting thing I notice is that if I use syscall.Rawsyscall(_exit), then the `close(done)` will not have effect as the runtime doesn't have a chance to schedule, unless I add `runtime.Gosched()` before the Rawsyscall.